project-koku/koku
# AWS org unit feature

Connected issues: 
* Add AWS organizations API https://github.com/project-koku/koku/issues/1916
* AWS Organizational unit crawler should also record delete timestamps https://github.com/project-koku/koku/issues/2043
* Correlate AWS costs with Organization Unit https://github.com/project-koku/koku/issues/1915
* Add group by and filter option for org unit to AWS, OCP on AWS reports options https://github.com/project-koku/koku/issues/1917 - this is only implemented for the AWS reports in this PR. 

Marking WIP until a couple of cleanup items & questions are taken care of! Feel free to give feedback at this time.

---- 
# How to test
1. New `group_by[org_unit_id]` & `filter[org_unit_id]` for the aws reports endpoint.
We have added a make command that will populate a aws organizational data tree inside of the database for you. Therefore all you have to do is make sure that the server is up and ready and run:
```
 make load-aws-org-unit-tree
```
After the org tree has been upload you can now check the new group by/filter with the following links:
http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?group_by[org_unit_id]=R_001
http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?filter[org_unit_id]=OU_002



2. The new organizations endpoint. 
After running the `make load-aws-org-unit-tree` the data in the organizations endpoint will also be populated. So can view this data through the following links:
- http://127.0.0.1:8000/api/cost-management/v1/organizations/aws/?key_only=True
- http://127.0.0.1:8000/api/cost-management/v1/organizations/aws/

**Note:** The organizations api is disassociated with the costs. Therefore, if you are seeing an account in the org api, but not the cost api it likely means that the account did not have any cost associated with it. 



3. The org unit crawler. 
Testing the crawler is a little bit more complicated because you will need a real aws account with aws organizational units in it. If you do not have one, message me or @myersCody and we can provide you with one. 

After you have an aws account you will then need to manually change the crawler to run more frequently than once a day:
```
app.conf.beat_schedule["crawl_account_hierarchy"] = {
    "task": "masu.celery.tasks.crawl_account_hierarchy",
    "schedule": crontab(minute="*"),
}
```
You may still want to be careful here because when you start the process a download task may begin to run. If you are using a large account like we were  while developing this you will for sure not want to download these massive files. Therefore you may want to stop the downloading with a 
```
docker-compose down && docker-compose up -d
```
If at any point, if anyone has questions feel free to reach out to @abaiken
